There is a apyori file which is attached in template directory which has been provided by python software foundation. 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
#importing dataset without header
dataset = pd.read_csv('Dataset\\Market_Basket_Optimisation.csv')
dataset.head(2)

shrimp    almonds avocado vegetables mix green grapes whole weat flour  \
0  burgers  meatballs    eggs            NaN          NaN              NaN   
1  chutney        NaN     NaN            NaN          NaN              NaN   

  yams cottage cheese energy drink tomato juice low fat yogurt green tea  \
0  NaN            NaN          NaN          NaN            NaN       NaN   
1  NaN            NaN          NaN          NaN            NaN       NaN   

  honey salad mineral water salmon antioxydant juice frozen smoothie spinach  \
0   NaN   NaN           NaN    NaN               NaN             NaN     NaN   
1   NaN   NaN           NaN    NaN               NaN             NaN     NaN   

   olive oil  
0        NaN  
1        NaN

In [3]:
#importing the dataset with header
dataset = pd.read_csv('Dataset\\Market_Basket_Optimisation.csv',header=None)
dataset.head(2)

0          1        2               3             4   \
0   shrimp    almonds  avocado  vegetables mix  green grapes   
1  burgers  meatballs     eggs             NaN           NaN   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN

First column of the dataset contains different products sold at the grocerry store.<br>
Here there is a small misinterpretation that the first row is being considered as labeled column of the dataset which is not the case.

In [4]:
dataset.tail(2)

0                1            2               3    4    5    6   \
7499  escalope        green tea          NaN             NaN  NaN  NaN  NaN   
7500      eggs  frozen smoothie  yogurt cake  low fat yogurt  NaN  NaN  NaN   

       7    8    9    10   11   12   13   14   15   16   17   18   19  
7499  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
7500  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN

Apriori algorithm expects a list of list for different transactions.<br>
There will one big list containing all the different transactions and each transaction is going to be a list itself which will be name of products.

In [5]:
#Preparing the data for apriori algorithm
#last index value is 7501
transactions = []
for i in range(0, 7501):
    transactions.append([str(dataset.values[i,j]) for j in range(0, 20)])
#product will be a string inside the ' '

In [6]:
# Training Apriori on the dataset
from Template.apyori import apriori
rules = apriori(transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3, min_length = 2)

Values of the parameters being passed<br>
Until we are satisfied with the values of these parameters we will choose/try different values of the parameters.<br>
`min_length` = Minimum no. of products required in the basket.<br> 
`min_support` = The support of set of items `i` = the number of transctions containing the set of items `i ` divided by the total number of transactions.So the items thata are going to be appear in rule have highere support than the minimum support.<br>
`confidence` = If we set a high confidence then rules has to be correct upto that extent. This may rise a case that a `product is bought together` `not` because they are well `associated` but simply because each of the product is one of the most `purchased` product.<br>

In [7]:
#Let our goal is set such that the item is bought 3 or 4 times a day.
#So we will consder that a product is purchased 3 times a day. 
#In a week it is bought 3*7 = 21 times a week.
print('Support = ',(3*7)/7501)

Support =  0.0027996267164378083


In [8]:
# Visualising the results
results = list(rules)

In [9]:
results[:5]

[RelationRecord(items=frozenset({'light cream', 'chicken'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)]),
 RelationRecord(items=frozenset({'escalope', 'mushroom cream sauce'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)]),
 RelationRecord(items=frozenset({'pasta', 'escalope'}), support=0.005865884548726837, ordered_statistics=[OrderedStatistic(items_base=frozenset({'pasta'}), items_add=frozenset({'escalope'}), confidence=0.3728813559322034, lift=4.700811850163794)]),
 RelationRecord(items=frozenset({'honey', 'fromage blanc'}), support=0.003332888948140248, ordered_statistics=[OrderedStatistic(items_base=frozenset({'fromage blanc'}), items_add=frozenset({'honey'}), confidence=0

***
The above are the top 5 tresults.

**Analysis**<br>
***
For `first` combination : Items :- light cream and chicken<br>
Support = $\frac{light cream + chicken}{Total Products}$ = 0.004532728969470737 = 0.4%<br>
confidence = 0.29059829059829057 = 29% i.e. if people buy light cream then they have 29% chance that they will buy chicken.<br>
lift = 4.84395061728395 = 4.8<br>
***
For `Second` combination : Items :- mushroom cream sauce and escalope<br>
Support = $\frac{mushroom cream sauce + escalope}{Total Products}$ = 0.005732568990801226 = 0.5%<br>
confidence = 0.3006993006993007 = 30% i.e. if people buy mushroom cream sauce then they have 30% chance that they will buy escalope.<br>
lift = 3.790832696715049 = 3.8<br>
***
For `Fifth` combination : Items :- ground beef and herb & pepper<br>
Support = $\frac{ground beef + herb  pepper}{Total Products}$ = 0.015997866951073192 = 1.6%<br>
confidence = 0.3234501347708895 = 32% i.e. if people buy ground beef then they have 32% chance that they will buy herb & pepper.<br>
lift = 3.2919938411349285 = 3.2<br>
